src/controller/ReadInteraction.h

```diff
@@ -56,12 +56,63 @@ CHIP_ERROR ReadAttribute(Messaging::ExchangeManager * aExchangeMgr, const Sessio
     readParams.mpAttributePathParamsList    = &attributePath;
     readParams.mAttributePathParamsListSize = 1;
 
-    auto onDone = [](app::ReadClient * apReadClient, TypedReadCallback<DecodableAttributeType> * callback) {
+    auto onDone = [](app::ReadClient * apReadClient, TypedReadAttributeCallback<DecodableAttributeType> * callback) {
         chip::Platform::Delete(callback);
     };
 
-    auto callback = chip::Platform::MakeUnique<TypedReadCallback<DecodableAttributeType>>(clusterId, attributeId, onSuccessCb,
-                                                                                          onErrorCb, onDone);
+    auto callback = chip::Platform::MakeUnique<TypedReadAttributeCallback<DecodableAttributeType>>(clusterId, attributeId,
+                                                                                                   onSuccessCb, onErrorCb, onDone);
+    VerifyOrReturnError(callback != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    ReturnErrorOnFailure(engine->NewReadClient(&readClient, app::ReadClient::InteractionType::Read, callback.get()));
+
+    err = readClient->SendReadRequest(readParams);
+    if (err != CHIP_NO_ERROR)
+    {
+        readClient->Shutdown();
+        return err;
+    }
+
+    //
+    // At this point, we'll get a callback through the OnDone callback above regardless of success or failure
+    // of the read operation to permit us to free up the callback object. So, release ownership of the callback
+    // object now to prevent it from being reclaimed at the end of this scoped block.
+    //
+    callback.release();
+    return err;
+}
+
+/*
+ * A typed read event function that takes as input a template parameter that encapsulates the type information
+ * for a given attribute as well as callbacks for success and failure and either returns a decoded cluster-object representation
+ * of the requested attribute through the provided success callback or calls the provided failure callback.
+ *
+ * The EventTypeInfo is generally expected to be a ClusterName::Events::EventName::TypeInfo struct, but any
```

Comment:
There is no such struct.  What we have is `ClusterName::Events::EventName::Type` and `ClusterName::Events::EventName::DecodableType`.

Presumably this should use `DecodableType` and the template parameter should be `DecodableEventType` or so.